config/config_handler.go

```diff
@@ -0,0 +1,214 @@
+
+const (
+	pdConfHandlerRefreshInterval = 30 * time.Second
+	tidbComponentName            = "tidb"
+)
+
+type pdConfHandler struct {
+	id         string // ip:port
+	version    *configpb.Version
+	curConf    atomic.Value
+	interval   time.Duration
+	wg         sync.WaitGroup
+	exit       chan struct{}
+	pdConfCli  pd.ConfigClient
+	reloadFunc func(oldConf, newConf *Config)
+}
+
+func newPDConfHandler(localConf *Config, reloadFunc ConfReloadFunc,
+	newPDCliFunc func([]string, pd.SecurityOption) (pd.ConfigClient, error), // for test
+) (*pdConfHandler, error) {
+	addresses, _, err := ParsePath(localConf.Path)
+	if err != nil {
+		return nil, err
+	}
+	host := localConf.Host
+	if localConf.AdvertiseAddress != "" {
+		host = localConf.AdvertiseAddress
+	}
+	id := fmt.Sprintf("%v:%v", host, localConf.Port)
+	security := localConf.Security
+	if newPDCliFunc == nil {
+		newPDCliFunc = pd.NewConfigClient
+	}
+	pdCli, err := newPDCliFunc(addresses, pd.SecurityOption{
+		CAPath:   security.ClusterSSLCA,
+		CertPath: security.ClusterSSLCert,
+		KeyPath:  security.ClusterSSLKey,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	confContent, err := encodeConfig(localConf)
+	if err != nil {
+		return nil, err
+	}
+
+	// register to PD and get the new default config.
+	// see https://github.com/pingcap/tidb/pull/13660 for more details.
+	// suppose port and security config items cannot be change online.
+	status, version, conf, err := pdCli.Create(context.Background(), new(configpb.Version), tidbComponentName, id, confContent)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if status.Code != configpb.StatusCode_OK {
+		return nil, errors.New(fmt.Sprintf("fail to register config to PD, errmsg=%v", status.Message))
+	}
+
+	newConf, err := decodeConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+	if err := newConf.Valid(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ch := &pdConfHandler{
+		id:         id,
+		version:    version,
+		interval:   pdConfHandlerRefreshInterval,
+		exit:       make(chan struct{}),
+		pdConfCli:  pdCli,
+		reloadFunc: reloadFunc,
+	}
+	ch.curConf.Store(newConf)
+	return ch, nil
+}
+
+func (ch *pdConfHandler) Start() {
+	ch.wg.Add(1)
+	go ch.run()
+}
+
+func (ch *pdConfHandler) Close() {
+	close(ch.exit)
+	ch.wg.Wait()
+	ch.pdConfCli.Close()
+}
+
+func (ch *pdConfHandler) GetConfig() *Config {
+	return ch.curConf.Load().(*Config)
+}
+
+func (ch *pdConfHandler) run() {
+	defer func() {
+		if r := recover(); r != nil {
+			logutil.Logger(context.Background()).Error("[PDConfHandler] panic: " + fmt.Sprintf("%v", r))
+		}
+		ch.wg.Done()
+	}()
+	for {
+		select {
+		case <-time.After(ch.interval):
+			// fetch new config from PD
+			status, version, newConfContent, err := ch.pdConfCli.Get(context.Background(), ch.version, tidbComponentName, ch.id)
+			if err != nil {
+				logutil.Logger(context.Background()).Error("[PDConfHandler] fetch new config error", zap.Error(err))
```

Comment:
I think we do not need `[PDConfHandler]`

`Error("pdConfHandler fetches new config fail", zap.Error(err))` is enough